--- Heavy.pm	2015-01-25 01:14:31.000000000 +1100
+++ /c/strawberry/perl/site/lib/PAR/Heavy.pm	2016-08-05 10:52:14.000000000 +1000
@@ -145,10 +145,17 @@
     else {
         $filename = File::Spec->catfile(
             ($ENV{PAR_TEMP} || File::Spec->tmpdir),
-            ($name || ($member->crc32String . ".$DynaLoader::dl_dlext"))
+            #($name || ($member->crc32String . ".$DynaLoader::dl_dlext"))
+            ($name || ((File::Spec->splitpath ($member->fileName))[-1]))
         );
         ($filename) = $filename =~ /^([\x20-\xff]+)$/;
 
+### START HACK
+if (-e $filename && not $filename =~ /Glib|Gtk3|Gnome|Pango|Cairo/) {
+    $filename .= $member->crc32String; #  kludge workaround
+}
+### END HACK
+
         open $fh, '>', $filename or die $!
             unless -r $filename and -e _
                 and -s _ == $member->uncompressedSize;
