--- myldr/Makefile.PL	2012-01-27 05:51:55.000000000 +1100
+++ myldr/Makefile.PL	2012-10-30 16:36:48.437962500 +1100
@@ -115,8 +115,8 @@
     if( $^O =~ /^(?:MSWin|cygwin)/ ){
    $res = 'ppresource.coff';
    $pre_res = qq(winres\\pp.res);
-   $rt_cmd = qq(windres -F pei-i386 -i winres\\pp.rc -o $pre_res);
-   $res_cmd = qq(windres -o ppresource.coff $pre_res);
+   $rt_cmd = qq(windres -o $res -i winres\\pp.rc);
+   $res_cmd = qq(-$(NOOP));
    $res_section = $res;
     }
     $mt_cmd = '-$(NOOP)';
